src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,72 +245,48 @@ void addArg(RefKind refKind, BoundExpression expression)
-                        {
-                            // array[int]
-                            Debug.Assert(input.Type?.IsSZArray() == true);
-                            BoundExpression access = _factory.ArrayAccess(input, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerSymbol: PropertySymbol indexer } e:
-                        {
-                            // this[int]
-                            BoundExpression access = _factory.Indexer(input, indexer, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerAccess: BoundIndexerAccess indexerAccess } e:
+                    case BoundDagIndexerEvaluation e:
                         {
                             // this[Index]
-                            BoundExpression access = makeIndexerAccess(e.Syntax, indexerAccess, makeExplicitIndexArgument(e.Index, e.Index < 0));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
+                            // this[int]
+                            // array[Index]
+                            _localRewriter.AddPlaceholderReplacement(e.ReceiverPlaceholder, input);
+                            _localRewriter.AddPlaceholderReplacement<BoundListPatternUnloweredIndexPlaceholder>(e.ArgumentPlaceholder, makeUnloweredIndexArgument(e.Index));
 
-                    case BoundDagSliceEvaluation { IndexerAccess: null, SliceMethod: null } e:
-                        {
-                            // RuntimeHelpers.GetSubArray(T[], Range)
-                            TypeSymbol inputType = input.Type;
-                            Debug.Assert(inputType is { });
-                            Debug.Assert(inputType.IsSZArray());
-                            Debug.Assert(e.StartIndex >= 0 && e.EndIndex <= 0);
+                            var indexerAccess = e.IndexerAccess;
+                            if (indexerAccess is BoundIndexOrRangePatternIndexerAccess implicitAccess)
+                            {
+                                indexerAccess = implicitAccess.WithLengthOrCountAccess(_tempAllocator.GetTemp(e.LengthTemp));
```

Comment:
>_tempAllocator.GetTemp(e.LengthTemp) [](http://example.com/codeflow?start=87&length=36)

Is it Ok to allocate this temp if length isn't used?

Reply:
The only time a list-pattern doesn't use the length is `[.. <pattern>]`. In that scenario we do more work than we should (we compute length when not needed). Given that it's a edge case (why would one do that?), I don't think we should care.